crates/sui-core/src/authority/authority_store_pruner.rs

```diff
@@ -189,53 +105,51 @@ impl AuthorityStorePruner {
             if effects.iter().any(|effect| effect.is_none()) {
                 return Err(anyhow::anyhow!("transaction effects data is missing"));
             }
-            Self::handle_checkpoint(effects.into_iter().flatten(), &perpetual_db.objects)?;
-            checkpoint_store.update_highest_pruned_checkpoint(&checkpoint)?;
-            pruned_seq_number += 1;
+            batch_effects.extend(effects.into_iter().flatten());
+
+            if batch_effects.len() >= MAX_TRANSACTIONS_IN_ONE_BATCH
+                || checkpoints_in_batch >= MAX_CHECKPOINTS_IN_ONE_BATCH
+            {
+                Self::handle_checkpoint(batch_effects, &perpetual_db.objects)?;
+                checkpoint_store
+                    .update_highest_pruned_checkpoint(checkpoint_number, checkpoint_digest)?;
+                batch_effects = vec![];
+                checkpoints_in_batch = 0;
+            }
+        }
+        if !batch_effects.is_empty() {
+            Self::handle_checkpoint(batch_effects, &perpetual_db.objects)?;
+            checkpoint_store
```

Comment:
Can we afford to update this at the end? What is the worse that happens? We crash before updating, and try to prune again? Is this a problem?

Reply:
pruning should be idempotent. We update as we go once we fill the batch or at the end of the loop if there are no more checkpoints to process in the current epoch